### PR TITLE
[core] Fix a corner case where QueryAllWorkerStates never return

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
In some extreme cases, when the worker becomes dead after getting the workers, the RPC request might never return.

This PR fixed this issue.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -922,6 +925,9 @@ void NodeManager::QueryAllWorkerStates(
           }
         });
   }
+  if(all_dead) {
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:928]:
```suggestion
  if (all_dead) {
```

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Wonder if we can also add timeout?

**Comment by iycheng** [comment]
> Wonder if we can also add timeout?

maybe it's ok. if the worker died, it'll return failure. otherwise, it'll just retry I think?

**Comment by rkooo567** [comment]
I think we don't have retry there. One potential problem is the worker being overloaded and never reply, which hangs the reply of the RPC. I don't think we need to handle this in this PR

**Pull request merged**